### ISSUE TITLE
Multiple `match.any` breaks stub behavior

### ISSUE THREAD

**Issue by michaloo** [sinonjs/sinon#1887]
**Describe the bug**
I want to configure stub behavior to ignore first two function arguments using `sinon.match.any` and use third matcher to decide which response to return.
It was working all good till version 6.1.3, then it started to behave unexpectedly.

**To Reproduce**
1. Create the simplest script like this:
```js
const assert = require("assert");
const sinon = require("sinon");

const stub = sinon.stub();

stub.withArgs(sinon.match.any, sinon.match.any, sinon.match("a")).returns("a");
stub.withArgs(sinon.match.any, sinon.match.any, sinon.match("b")).returns("b");

assert.equal(stub({}, [], "a"), "a");
assert.equal(stub({}, [], "b"), "b");
```
2. Run the test with node

**Expected behavior**
The script should run successfully, the first stub call should match the first `withArgs` definition returning `a`, the second call should return `b`.

**Actual behavior**
First stub call returns `b` so the result for not matching `withArgs`.
`AssertionError [ERR_ASSERTION]: 'b' == 'a'`

The second call returns `undefined`, so it looks like nothing was matched
`AssertionError [ERR_ASSERTION]: undefined == 'b'`

**Context (please complete the following information):**

* Library version: 6.1.3 and above
* Environment: node v8.11.3

It looks like using `sinon.match.any` twice is the problem. If there is one argument "placeholder" thing works as expected. It does not matter what is passed as the actual arguments values.

I've tried to look at 6.1.2 (latest working version) and 6.1.3 diff but could not find an obvious problem.
Any help would be appreciated :)

**Comment by mroderick** [comment]
Using @michaloo's test and [@fatso83's very handy `git bisect run` template](https://github.com/fatso83/git-bisect-scripts/blob/master/full-template.sh) I was quickly able to determine that the first bad commit is: 014aecff40025b74e1bb27edeaceb5d5beb3ff6d

**Comment by mroderick** [comment]
When we fix this issue, a test should be added to `issues-test.js` to ensure this remains fixed, even when https://github.com/sinonjs/referee/issues/29 is finished.

**Comment by mgred** [comment]
@michaloo I add the test case and it works fine now :sunglasses: Could you confirm for your specific case?

**Comment by mroderick** [comment]
I am going to claim that this was fixed in #1918, and am closing this.
